commit e7d05ac81534b25796b78ccc36f9065d03962294
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2019-03-12T15:10:56-07:00 (7 months ago)
    
    TRITON-1304 net-agent fails to properly handle admin NIC for CNs
    Reviewed by: Jason King <jason.king@joyent.com>

diff --git a/lib/nic-fsm.js b/lib/nic-fsm.js
index a128fda..984625a 100644
--- a/lib/nic-fsm.js
+++ b/lib/nic-fsm.js
@@ -28,6 +28,7 @@ var ANTI_SPOOF_FIELDS = [
 
 var LOCAL_FIELDS = [
     'belongs_to_type',
+    'belongs_to_uuid',
     'cn_uuid',
     'owner_uuid',
     'primary',
@@ -110,33 +111,33 @@ function getDifferences(fields, cur, old) {
  * depicted here, but are loops back into the same state usually).
  *
  *                       +---------+
- *                       | create  | -------------------------+
- *                       +---------+                          |
- *                 404 on 1st ^                               |
- *                    refresh |    setLocal()                 |
- *       +------+        +---------+ -----> +--------+        |
- *       | init | -----> | refresh |        | update |        |
- *       +------+        +---------+   +--- +--------+        |
- *                        |   ^        |         |            |
- *           NAPI 404 for |   |        |         |            |
- *           existing NIC |   |        |         |            |
- *       +---------+      |   |        |         |            |
- *       | remove  | <----+   |        |         |            |
- *       +---------+          |        |         |            |
- *            |               |        |         |            |
- *            v               |        |         v            |
- *       +---------+     +---------+   |    +--------------+  |
- *       | stopped |     | waiting | <-+--- | update.local | -+
- *       +---------+     +---------+   |    +--------------+  |
- *            ^                        |                      |
- *            |          +---------+   +--- +--------------+  |
- *            +----<---- | release |        | update.napi  | -+
- *            |          +---------+        +--------------+  |
- *            |               ^                  |            |
- *            |               |     NAPI 404     |            |
- *            +----<--------- | -----------------+            |
- *                            |        releaseFrom()          |
- *                            +-------------------------------+
+ *                       | create  | --------------------------+
+ *                       +---------+                           |
+ *                 404 on 1st ^                                |
+ *                    refresh |    setLocal()                  |
+ *       +------+        +---------+ -----> +--------+         |
+ *       | init | -----> | refresh |        | update |         |
+ *       +------+        +---------+   +--- +--------+---->----|--+
+ *                        |   ^        |         |             |  |
+ *           NAPI 404 for |   |        |         |             |  |
+ *           existing NIC |   |        |         |             |  |
+ *       +---------+      |   |        |         |             |  |
+ *       | remove  | <----+   |        |         |             |  |
+ *       +---------+          |        |         |             |  |
+ *            |               |        |         |             |  |
+ *            v               |        |         v             |  |
+ *       +---------+     +---------+   |    +--------------+   |  |
+ *       | stopped |     | waiting | <-+--- | update.local | --+  |
+ *       +---------+     +---------+   |    +--------------+   |  |
+ *            ^                        |                       |  |
+ *            |          +---------+   +--- +--------------+ <-|--+
+ *            +----<---- | release |        | update.napi  | --+
+ *            |          +---------+        +--------------+   |
+ *            |               ^                  |             |
+ *            |               |     NAPI 404     |             |
+ *            +----<--------- | -----------------+             |
+ *                            |        releaseFrom()           |
+ *                            +--------------------------------+
  */
 function NicFSM(opts) {
     assert.object(opts, 'opts');
@@ -318,6 +319,7 @@ NicFSM.prototype.state_update = function (S) {
         'refresh',
         'remove',
         'update.local',
+        'update.napi',
         'waiting'
     ]);
 
@@ -331,6 +333,27 @@ NicFSM.prototype.state_update = function (S) {
         return;
     }
 
+    /*
+     * When a new CN is first booted, booter inserts a NAPI record for the admin
+     * NIC which has:
+     *
+     *   belongs_to_type = other
+     *   belongs_to_uuid = <admin user's UUID>
+     *
+     * if we see this for a NIC we "own", we should update NAPI so the type gets
+     * corrected.
+     */
+    if (this.remote.belongs_to_type === 'other' &&
+        this.remote.belongs_to_uuid === this.app.admin_uuid &&
+        this.local.belongs_to_type === 'server') {
+        this.log.info({
+            nic: this.local
+        }, 'found our NIC unclaimed in NAPI, claiming.');
+
+        S.gotoState('update.napi');
+        return;
+    }
+
     /*
      * If the belongs_to_uuid has changed, then we need to move the NIC.
      */
diff --git a/package.json b/package.json
index c35a294..340cfd7 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "net-agent",
     "description": "Triton Network Agent",
-    "version": "2.2.4",
+    "version": "2.2.5",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
